pkg/streamingpromql/aggregations/common.go

```diff
@@ -0,0 +1,26 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package aggregations
+
+import (
+	"github.com/prometheus/prometheus/promql/parser"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/functions"
+	"github.com/grafana/mimir/pkg/streamingpromql/limiting"
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+)
+
+// AggregationFunction accumulates series that have been grouped together and computes the output series data.
+type AggregationFunction interface {
+	// AccumulateSeries takes in a series as part of the group
+	AccumulateSeries(data types.InstantVectorSeriesData, steps int, start int64, interval int64, memoryConsumptionTracker *limiting.MemoryConsumptionTracker, emitAnnotationFunc functions.EmitAnnotationFunc) error
+	// ComputeOutputSeries does any final calculations and returns the grouped series data
+	ComputeOutputSeries(start int64, interval int64, memoryConsumptionTracker *limiting.MemoryConsumptionTracker) (types.InstantVectorSeriesData, bool, error)
+}
+
+type AggregationFunctionFactory func() AggregationFunction
+
+var AggregationFunctionFactories = map[parser.ItemType]AggregationFunctionFactory{
+	parser.MAX: func() AggregationFunction { return &MaxAggregationFunction{} },
+	parser.SUM: func() AggregationFunction { return &SumAggregationFunction{} },
```

Comment:
It'd be interesting to benchmark the performance impact of this, but previously we were pooling `group` instances, and I imagine we'll want to do the same with these `AggregationFunction` instances too, with a pool per type.

In fact, we could change `AggregationFunctionFactories` to be something like `map[parser.ItemType]somePoolType` that `Aggregation` can use.

Reply:
We didn't see a huge jump in the benchmarks for creating the AggreationFunctionFactories, so I'll leave that for now. The jump in the benchmark was attributed to using `emitAnnotation` instead of `emitAnnotationFunc`, so with that fixed I expect the next benchmark run to be better. I'll mark this as resolved for now and I'll redo benchmarks later and we can pick up there if we've lost any efficiencies.